### DISCUSSION TITLE
How to delete the results of a Multi Map - Reduce index?

### DISCUSSION THREAD

**Discussion by dtaalbers** [ravendb/ravendb#14496] [Q&A] [answered]
I've got a multimap reduce index (added below) that fetches information from multiple collections and does some logic in the reduce that decides whether or not an item is obsolete and if it should be deleted. When I query the index it returns the correct information. But when I try to delete the results of the query via a patch I get the following error.

```
Cannot execute bulk operation on Map-Reduce indexes.
```

I am using the following patch (in RavenDB studio) when trying to delete the documents:

```
from index 'StockItems/GetObsolete' as s
where s.IsConfigured = false and s.Sku in ('207275-G', 'GGOEAFKA087499')
update {
    del(s.Id);
}
````
 How can I best solve this? 

Index:
```
public class StockItems_GetObsolete : AbstractMultiMapIndexCreationTask<StockItems_GetObsolete.Result>
{
    public class Result
    {
        public string Id { get; set; }
        public string Sku { get; set; }
        public List<string> ConfiguredWarehouses { get; set; }
        public string WarehouseId { get; set; }
        public bool IsConfigured { get; set; }
        public string Type { get; set; }
    }

    public StockItems_GetObsolete()
    {
        AddMap<Product>(
            products => from product in products
                        where product.Type == ProductTypes.Simple
                        where product.Variants.Count == 0
                        select new
                        {
                            product.Id,
                            product.Sku,
                            ConfiguredWarehouses = product.StockLimitItems.Select(x => x.Warehouse.Id).ToList(),
                            WarehouseId = (string) null,
                            IsConfigured = false,
                            Type = "Product",
                        }
        );

        AddMap<Product>(
            products => from product in products
                        where product.Type == ProductTypes.Simple
                        where product.Variants.Count > 0
                        from variant in product.Variants
                        select new
                        {
                            product.Id,
                            variant.Sku,
                            ConfiguredWarehouses = variant.StockLimitItems.Select(x => x.Warehouse.Id).ToList(),
                            WarehouseId = (string) null,
                            IsConfigured = false,
                            Type = "Variant"
                        }
        );

        AddMap<StockItem>(
            items => from item in items
                     select new
                     {
                         item.Id,
                         item.Sku,
                         ConfiguredWarehouses = (List<string>) null,
                         WarehouseId = item.Warehouse.Id,
                         IsConfigured = false,
                         Type = "StockItem"
                     }
        );

        Reduce = results => from result in results
                            group result by result.Sku
                            into grouped
                            let variant = grouped.SingleOrDefault(x => x.Type == "Variant")
                            let product = grouped.SingleOrDefault(x => x.Type == "Product")
                            let configuredWarehouses = variant != null ? variant.ConfiguredWarehouses : product.ConfiguredWarehouses
                            from item in grouped.Where(x => x.Type == "StockItem")
                            select new
                            {
                                Id = item.Id,
                                Sku = grouped.Key,
                                ConfiguredWarehouses = configuredWarehouses,
                                item.WarehouseId,
                                IsConfigured = configuredWarehouses.Contains(item.WarehouseId),
                                item.Type
                            };

        Indexes.Add(x => x.Sku, FieldIndexing.Default);
    }
}
```

**Comment by ayende** [comment, marked as answer]
Use an output collection and operate on that is likely the best option.

**Reply by ayende** [reply]
https://issues.hibernatingrhinos.com/issue/RavenDB-18897

**Reply by dtaalbers** [reply]
Understood. Would prefer to do it on the collection itself though so I will keep an eye on the opened issue. Thank you!

**Reply by dtaalbers** [reply]
I've played around with the output collection option. Whilst saving the output to a collection works fine, it often lacks behind.

1) I delete a warehouse from the product configuration and update the product (SaveChangesAsync) which triggers the index to check for obsolete items. 
2) I patch the collection `StockItems/Obsolete` right away with the following patch:
```
from "StockItems/Obsolete" as s
update {
    del(s.Id);
}
```

The obsolete stock item is added to the `StockItems/Obsolete` collection, but it is added after the patch already finished executing. Resulting in the `StockItems/Obsolete` collection still holding the obsolete stock item. I think this issue is solved when I am able to do a patch query on the map/reduce index. 🙏

**Reply by garayx** [reply]
you might want to wait for indexes to become non stale before you send the patch,
https://ravendb.net/docs/article-page/5.3/csharp/client-api/session/saving-changes#waiting-for-indexes

**Reply by dtaalbers** [reply]
> you might want to wait for indexes to become non stale before you send the patch, https://ravendb.net/docs/article-page/5.3/csharp/client-api/session/saving-changes#waiting-for-indexes

Yes, of course. That did the trick. Thank you very much! Still prefer to patch the index results though 😀.

**Reply by ayende** [reply]
We'll probably add that support

**Reply by dtaalbers** [reply]
> We'll probably add that support

Sounds good! Thanks for everything. I'll keep an eye out for a possible update.

**Reply by georgiosd** [reply]
@dtaalbers You could also use a subscription on the output collection to avoid having to wait for the index yourself. It will just trigger automatically.

**Comment by mcrio** [comment]
Hi. I'm writing a RavenDB storage library for OpenIdDict. In order to delete outdated data I need to rely on a Multi-Map Reduce index.

At this point in time, is there any way to do that without having to create an index output collection?

**Reply by ayende** [reply]
There is no way to run `DeleteByQuery` or `UpdateByQuery` on a map/reduce index.

**Reply by mcrio** [reply]
Thanks @ayende   As there is an open ticket  https://issues.hibernatingrhinos.com/issue/RavenDB-18897  do you think this will be supported in near future?

**Reply by ppekrol** [reply]
Apologies, but we do not have any immediate plans to support this is upcoming months.
